README.md
# PDS Feedback Widget

PDS Feedback Widget Javascript-based overlay and modal window that provides a user with a form to provide comments or feedback on the web page they are currently accessing. When a comment is submitted, it will be sent to pds_operator@jpl.nasa.gov and forwarded to the applicable node that has the application deployed.


## Register

Before being able to use this widget, you will need to register your hostname with the Engineering Node in order to enable the Google Re-captcha that is part of the application.

Email pds_operator@jpl.nasa.gov with the following information:
* Hostname(s) for your website
* Point(s) of contact to receive the Feedback


## Install

Here are the steps for deploying the Feedback widget to your website on an Apache Web Server:

>> NOTE: These installation instructions assume the use of an Apache Web Server. The Feedback widget can be installed on other web servers, but the exact steps may differ depending upon the software and configuration.


1. [Register](#register) with Engineering Node if you have not already.

2. Add the following code snippet between the `<head>` tag on each page of your website. If your website uses a header file, you can place it in there:
```
<!-- PDS Feedback Widget -->
<!-- Only add JQuery if you do not already include a library -->
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
<script src='https://www.google.com/recaptcha/api.js?onload=onloadCallback&render=explicit' async defer></script>
<link rel="stylesheet" href="/feedback/css/feedback.css"  type="text/css" media="screen" />
<script src="/feedback/js/modernizr-custom.js"></script>
<script src="/feedback/js/config.js"></script>
<script src="/feedback/js/feedback.js"></script>
<!-- -->
```

3. Download the Feedback Widget tar or zip from [Github](https://github.com/NASA-PDS-Incubator/feedback-widget/releases/latest) to your local machine, then copy to your home directory on the machine hosting the website.

4. From your home directory on the machine hosting your website, unpack the widget:
```
tar -xvzf <.tar.gz>

#or

unzip <.zip>
```

5. Next, configure the Feedback Widget . From the root folder of the widget, open `feedback/js/config.js` in your text editor of choice.<br><br>
A list and explanation of optional configurable variables can be found in the [following section](#configurable-variables). *Note: It is highly recommended that you update **additionalLinks** to provide help information unique to your website.*

6. Now we want to move the files needed for the widget to the home directory for your website, or *WEB_HOME_PATH*. This *WEB_HOME_PATH* is the path where your homepage resides. For example, if your homepage is `/my/website/index.html`, then your *WEB_HOME_PATH* is `/my/website/`. To move the files, from the command-line:

   * *macOS / Linux*
     ```
     # Make sure you are in the the feedback widget
     $ pwd
     /path/to/feedback-widget

     # Run deploy.sh to push the files to WEB_HOME_PATH
     $ ./deploy.sh /my/website
     sending incremental file list
     feedback/
     feedback/recaptcha-v3-verify.php
     feedback/css/
     feedback/css/feedback.css
     feedback/image/
     feedback/image/iconfinder_Help.svg
     feedback/image/msg_icon.png
     feedback/js/
     feedback/js/config.js
     feedback/js/feedback.js 
     feedback/js/modernizr-custom.js  
     ```

   * *Windows*
     TBD

## Configurable Variables

**Variable**    | **Description**                               | **Default Value**         | **Accepted Values or Types**<sup>[0](#zero)</sup>
--------------- | --------------------------------------------- | ------------------------- | --------------------
header          | The header or title in the Feedback popup.    | "Help Desk"               | text
feedbackType    | Type of feedback for the user to specify if he or she so desires.| "Comment,Question,Problem/Bug,Kudos,Other" | text<sup>[1](#first)</sup>
additionalLinks | Additional link(s) the user may find helpful. | ---                       | ---
&emsp;- title   | The text displayed for the link.              | "Information for Data Users"<br>"Information for Proposers"<br>"Information for Providers"<br>"OpenPlanetary Forum"<br>"Site Map" | text
&emsp;- url     | The URL of the link.                          | "https://pds.nasa.gov/home/users/"<br>"https://pds.nasa.gov/home/proposers/"<br>"https://pds.nasa.gov/home/providers/"<br>"https://forum.openplanetary.org"<br>"https://pds.nasa.gov/site-help.shtml" | valid URLs
|               |                                               |                           |
label           | The text on the Feedback tab.                 | "Need Help?"              | text
color           | The color of the Feedback tab.                | "#0b3d91" (NASA blue)     | text<sup>[2](#second)</sup>
fontColor       | The color of the text on the Feedback tab.    | "#ffffff" (white)         | text<sup>[2](#second)</sup>
fontSize        | The size of the text on the Feedback tab.     | "16"                      | integer<sup>[3](#third),[5](#fifth)</sup>
size            | ---                                           | ---                       | ---
&emsp;- width   | The width of the Feedback tab.                | "150"                     | integer<sup>[3](#third),[5](#fifth)</sup>
&emsp;- height  | The height of the Feedback tab.               | "60"                      | integer<sup>[3](#third),[5](#fifth)</sup>
placement       | ---                                           | ---                       | ---
&emsp;- side    | The side of screen to attach the Feedback tab.| "right"                   | "RIGHT" "LEFT" "TOP" "BOTTOM"
&emsp;- offset  | The offset from top or left side of the screen.<sup>[7](#seventh)</sup>|"50"| integer<sup>[4](#fourth),[5](#fifth)</sup> &#8712; [0,100]<sup>[6](#sixth)</sup>

<a name="zero"></a><sup>0:</sup> All values must be enclosed within double quotation marks.<br>
<a name="first"></a><sup>1:</sup> Separate multiple values with a comma.<br>
<a name="second"></a><sup>2:</sup> Must be written as a hexadecimal, RGB, or HSL color, or be from [this list](https://www.w3schools.com/colors/colors_names.asp) of accepted color names. If you are unfamiliar with these formats, sites such as [ColorHexa](https://www.colorhexa.com) can help you find an exact color in the accepted formats. Be sure to include '#', 'rgb()', or 'hsl()' within the double quotations in config.js as needed.<br>
<a name="third"></a><sup>3:</sup> Used with the unit 'px' at 1px = 1/96th of an inch. For a digital facsimile of how different px values render on your screen, there are resources such as this [w3schools widget](https://www.w3schools.com/cssref/tryit.asp?filename=trycss_unit_px). (To use, adjust the integer value for any font-size in the left pane then click the green 'Run' button near the top to refresh the results in the right pane.)<br>
<a name="fourth"></a><sup>4:</sup> Used with the unit 'vh' or 'vw' at 1vh or 1vw = 1/100th of the browser window's height or width.<br>
<a name="fifth"></a><sup>5:</sup> Enter the integer value only. Do not include such characters as '%', 'vh', 'vw', 'px', 'pt', 'em', 'rem', etc.<br>
<a name="sixth"></a><sup>6:</sup> If this value is not in the range of 0 to 100 (inclusive) or is such that a part of the tab would not be visible, the most extreme value allowed which keeps the entire tab visible will be applied.<br>
<a name="seventh"></a><sup>7:</sup> Depending on the side of screen specified, this places the vertical center of the tab proportional to the height of the screen or the left corner of the tab proportional to the width of the screen. For instance, if you fill in config.js as follows:
  ```
    placement: {
      side: "bottom",
      offset: "0"
    }
  ```
  the Feedback tab will be on the bottom of the screen squarely in the left corner.
  ```
    placement: {
      side: "left",
      offset: "25"
    }
  ```
  places the Feedback tab on the left side of the screen a quarter of the way down from the top.<br>


## Uninstall

The following steps will help you uninstall the feedback widget:

1. Remove the following code snippet from any web pages it was added to:

```
<!-- PDS Feedback Widget -->
<!-- Only add JQuery if you do not already include a librar -->
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
<script src='https://www.google.com/recaptcha/api.js?render=explicit' async defer></script>
<link rel="stylesheet" href="/feedback/css/feedback.css"  type="text/css" media="screen" />
<script src="/feedback/js/modernizr-custom.js"><script>
<script src="/feedback/js/config.js"></script>
<script src="/feedback/js/feedback.js"></script>
<!-- -->
```

2. Remove the files from *WEB_HOME_PATH/feedback*:
```
# Unix example
rm -fr /my/website/feedback
```

## License
[LICENSE.txt](LICENSE.txt)


CHANGELOG.md
# Change Log

## [v1.2.0](https://github.com/NASA-PDS-Incubator/feedback-widget/tree/v1.2.0) (2019-09-09)
[Full Changelog](https://github.com/NASA-PDS-Incubator/feedback-widget/compare/v1.1.1...v1.2.0)

**Fixed bugs:**

- Unable to change font size from config.js [\#7](https://github.com/NASA-PDS-Incubator/feedback-widget/issues/7)

**Closed issues:**

- Readme edits [\#5](https://github.com/NASA-PDS-Incubator/feedback-widget/issues/5)

**Merged pull requests:**

- Address Issue \#7 and misc other improvements [\#8](https://github.com/NASA-PDS-Incubator/feedback-widget/pull/8) ([c-suh](https://github.com/c-suh))

## [v1.1.1](https://github.com/NASA-PDS-Incubator/feedback-widget/tree/v1.1.1) (2019-08-07)
[Full Changelog](https://github.com/NASA-PDS-Incubator/feedback-widget/compare/v1.1.0...v1.1.1)

**Merged pull requests:**

- Readme \(and consequent\) edits [\#6](https://github.com/NASA-PDS-Incubator/feedback-widget/pull/6) ([c-suh](https://github.com/c-suh))
- Merge dev into master [\#4](https://github.com/NASA-PDS-Incubator/feedback-widget/pull/4) ([jordanpadams](https://github.com/jordanpadams))
- Prep for release v1.1.0 [\#3](https://github.com/NASA-PDS-Incubator/feedback-widget/pull/3) ([jordanpadams](https://github.com/jordanpadams))

## [v1.1.0](https://github.com/NASA-PDS-Incubator/feedback-widget/tree/v1.1.0) (2019-06-19)
[Full Changelog](https://github.com/NASA-PDS-Incubator/feedback-widget/compare/v1.0...v1.1.0)

**Merged pull requests:**

- PDS-891: deprecate Feedback app by triggering new Feedback widget [\#2](https://github.com/NASA-PDS-Incubator/feedback-widget/pull/2) ([c-suh](https://github.com/c-suh))
- PDS-876: Implement configuration changes [\#1](https://github.com/NASA-PDS-Incubator/feedback-widget/pull/1) ([c-suh](https://github.com/c-suh))

## [v1.0](https://github.com/NASA-PDS-Incubator/feedback-widget/tree/v1.0) (2019-05-15)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

deploy.sh
#!/bin/bash
#############################################################################
# Copyright (c) 2019, California Institute of Technology ("Caltech").  
# U.S. Government sponsorship acknowledged.
#
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of source code must retain the above copyright notice,
#   this list of conditions and the following disclaimer.
# * Redistributions must reproduce the above copyright notice, this list of
#   conditions and the following disclaimer in the documentation and/or other
#   materials provided with the distribution.
# * Neither the name of Caltech nor its operating division, the Jet Propulsion
#   Laboratory, nor the names of its contributors may be used to endorse or
#   promote products derived from this software without specific prior written
#   permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#############################################################################
#
# deploy.sh
#
# Script to deploy applications to web server
#

DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
if [ -n "${1}" ] && [ -d "${1}" ]; then
    webhome=${1}
elif [ -z "${APACHE_HOME}" ]; then
    echo "ERROR: Path to Web Server directory containing website must be specified." 1>&2
    echo 1>&2
    echo "Usage: $(basename $0) <path/to/www/>" 1>&2
    echo
    exit 1
fi

rsync -av ${DIR}/feedback $webhome/

exit 0

feedback/js/feedback.js
/*
    Copyright (c) 2019, California Institute of Technology ("Caltech").
    U.S. Government sponsorship acknowledged.
    All rights reserved.
    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions are met:
    * Redistributions of source code must retain the above copyright notice,
      this list of conditions and the following disclaimer.
      * Redistributions must reproduce the above copyright notice, this list of
      conditions and the following disclaimer in the documentation and/or other
      materials provided with the distribution.
      * Neither the name of Caltech nor its operating division, the Jet Propulsion
      Laboratory, nor the names of its contributors may be used to endorse or
      promote products derived from this software without specific prior written
      permission.
    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
    ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
    POSSIBILITY OF SUCH DAMAGE.
*/

document.addEventListener("DOMContentLoaded", function(){
	Feedback(config);
});

(function( window, document, undefined ) {
	if ( window.Feedback !== undefined ) {
		return;
	}

	// log proxy function
	var log = function( msg ) {
		window.console.log( msg );
	},
	// function to remove elements, input as arrays
	removeElements = function( remove ) {
		for (var i = 0, len = remove.length; i < len; i++ ) {
			var item = Array.prototype.pop.call( remove );
			if ( item !== undefined ) {
				if (item.parentNode !== null ) { // check that the item was actually added to DOM
					item.parentNode.removeChild( item );
				}
			}
		}
	},
	loader = function() {
		var div = document.createElement("div"), i = 3;
		div.className = "feedback-loader";

		while (i--) { div.appendChild( document.createElement( "span" )); }
		return div;
	},
	getBounds = function( el ) {
		return el.getBoundingClientRect();
	},
	emptyElements = function( el ) {
		var item;
		while( (( item = el.firstChild ) !== null ? el.removeChild( item ) : false) ) {}
	},
	element = function( name, text ) {
		var el = document.createElement( name );
		el.appendChild( document.createTextNode( text ) );
		return el;
	},
	// script onload function to provide support for IE as well
	scriptLoader = function( script, func ){
		if (script.onload === undefined) {
			// IE lack of support for script onload

			if( script.onreadystatechange !== undefined ) {

				var intervalFunc = function() {
					if (script.readyState !== "loaded" && script.readyState !== "complete") {
						window.setTimeout( intervalFunc, 250 );
					} else {
						// it is loaded
						func();
					}
				};

				window.setTimeout( intervalFunc, 250 );

			} else {
				log("ERROR: We can't track when script is loaded");
			}

		} else {
			return func;
		}

	},
	sendButton,
	captchaUrl = "/feedback/recaptcha-v3-verify.php",
	feedbackUrl = "/email-service/SubmitFeedback",
	modal = document.createElement("div"),
	modalBody = document.createElement("div"),
	modalHeader = document.createElement("div"),
	modalFooter = document.createElement("div"),
	captchaScore = 0;

	window.captchaCallback = function( response ) {};
	
	window.Feedback = function( options ) {

		options = options || {};

		// default properties
		options.host = options.host || "";
		options.feedback.header = options.feedback.header || "Help Desk";
		options.page = options.page || new window.Feedback.Form();

		var glass = document.createElement("div"),
			returnMethods = {

			// open send feedback modal window
			open: function() {
				options.page.render();
				document.body.appendChild( glass );
				button.disabled = true;

				// modal close button
				var a = element("a", "x");
				a.className =  "feedback-close";
				a.onclick = returnMethods.close;
				a.href = "#";

				// build header element
				modalHeader.appendChild( a );
				modalHeader.appendChild( element("h3", options.feedback.header ) );
				modalHeader.className =  "feedback-header";

				modalBody.className = "feedback-body";

				emptyElements( modalBody );
				modalBody.appendChild( element("p", "How can we help you? Send us your question or feedback and we will get back to you within 1 business day.") );
				modalBody.appendChild( options.page.dom );
				var links = options.feedback.additionalLinks;
				if ( links !== "" ) {
					var additionalHelp = element("p", "In the meantime, you may find the following links helpful:"),
						additionalLinks = document.createElement("ul");
					additionalHelp.className = "additionalHelp";
					for (var i = 0; i < links.length; i++) {
						additionalLinks.insertAdjacentHTML('beforeend', '<li><a href="' + links[i].url + '">' + links[i].title + '</a></li>');
					}
					additionalHelp.insertAdjacentElement("beforeend", additionalLinks);
					modalBody.insertAdjacentElement("beforeend", additionalHelp);
					window.additionalHelp = additionalHelp;
				}
                                
				// Send button
				sendButton = document.createElement("input");
				sendButton.type = "submit";
				sendButton.value = "Send Feedback";
				sendButton.setAttribute("class", "feedback-btn g-recaptcha");
				sendButton.setAttribute("data-callback", "captchaCallback");
				sendButton.setAttribute("id", "recaptcha");

				modalFooter.className = "feedback-footer";
				modalFooter.appendChild( sendButton );

				modal.setAttribute("id", "feedback-form");
				modal.className = "feedback-modal";

				modal.appendChild( modalHeader );
				modal.appendChild( modalBody );
				modal.appendChild( modalFooter );

				document.body.appendChild( modal );

				window.grecaptcha.render("recaptcha", {sitekey: "6LfLCIgUAAAAAI3xLW5PQijxDyZcaUUlTyPDfYlZ"});
			},

			setupClose: function() {
				emptyElements( modalBody );
				sendButton.disabled = false;

				sendButton.value = "Close";

				sendButton.onclick = function() {
					returnMethods.close();
					return false;
				};
			},

			// close modal window
			close: function() {

				button.disabled = false;

				window.grecaptcha.reset();

				// remove feedback elements
				emptyElements( modalHeader );
				emptyElements( modalFooter );
				removeElements( [ modal, glass ] );

				return false;
			},

			// send data
			send: function( adapter ) {

				// make sure send adapter is of right prototype
				if ( !(adapter instanceof window.Feedback.Send) ) {
					throw new Error( "Adapter is not an instance of Feedback.Send" );
				}

				data = options.page.data();
				emptyElements( modalBody );
				modalBody.appendChild( loader() );

				// send data to adapter for processing
				adapter.send( data, function( success ) {

					returnMethods.setupClose();

					modalBody.setAttribute("class", "feedback-body confirmation");
					var message = document.createElement("p");
					if ( success === true ) {
						message.innerHTML = 'Thank you for making the PDS a better site.<br/>If you provided an email address, a PDS representative will get back to you as soon as possible.';
					} else {
						message.innerHTML = 'There was an error sending your feedback.<br/>If the problem persists, please email <a href="mailto:pds_operator@jpl.nasa.gov">pds_operator@jpl.nasa.gov</a>.';
					}
					modalBody.appendChild( message );

					if ( window.additionalHelp ) {
						modalBody.appendChild( window.additionalHelp );    
					}
				});

			},

			onloadCallback: function() {
				if ( new URLSearchParams(window.location.search).get("feedback") === "true" ) {
					returnMethods.open();
				}
			},

			captchaCallback: function( response ) {
				if ( document.getElementById("feedback-comment").reportValidity() ) {
					$.ajax({
						type: "POST",
						url: captchaUrl,
						data: {response: response},
						success: function (data) {
							//console.log(data);
							captchaScore = parseFloat(data.substring(data.indexOf("float") + 6, data.indexOf("float") + 9));
							if (captchaScore > 0.70) {
								options.url = options.url || options.host + feedbackUrl;
								options.adapter = options.adapter || new window.Feedback.XHR(options.url);
								emptyElements(modalBody);
								returnMethods.send(options.adapter);
							} else {
								modalBody.setAttribute("class", "feedback-body suspectedBot");
								document.getElementById("recaptcha").disabled = true;
								modalBody.insertAdjacentElement("afterbegin", element("p", "Are you a bot? Suspicious behavior detected."));
							}
						},
						error: function (XMLHttpRequest, textStatus, errorThrown) {
							modalBody.setAttribute("class", "feedback-body captchaError");
							returnMethods.setupClose();
							var message = document.createElement("p");
							message.innerHTML = '<b>Status: </b>' + textStatus + '; <b>Error: </b>' + errorThrown + '<br/>If the problem persists, please email <a href="mailto:pds_operator@jpl.nasa.gov">pds_operator@jpl.nasa.gov</a>.';
							modalBody.insertAdjacentElement("afterbegin", message);

							if ( window.additionalHelp ) {
								modalBody.appendChild( window.additionalHelp );
							}
						}
					});
				} else {
					return false;
				}
				window.grecaptcha.reset();
			}

		};

		window.onloadCallback = returnMethods.onloadCallback;
		window.captchaCallback = returnMethods.captchaCallback;

		glass.className = "feedback-glass";

		var button = document.createElement("button");

		if ( Modernizr.touchevents && window.screen.width < 1025 ) {
			var $window = $(window),
				docHeight = $(document).height(),
				rafId;
			window.requestAnimationFrame = window.requestAnimationFrame
				|| window.mozRequestAnimationFrame
				|| window.webkitRequestAnimationFrame
				|| window.msRequestAnimationFrame;
			$window.on("scroll", function() {
				if ( $window.scrollTop() + $window.height() > docHeight - 65 ) {
					rafId = window.requestAnimationFrame(function() {
						var offset = ($window.scrollTop() - 65) * ($window.scrollTop() - 65) * 0.00001;
						button.style.webkitTransform = "translateY(-" + offset + "px)";
						button.style.mozTransform = "translateY(-" + offset + "px)";
						button.style.transform = "translateY(-" + offset + "px)";
					});
				} else {
					window.cancelAnimationFrame(rafId);
					button.style.webkitTransform = "initial";
					button.style.mozTransform = "initial";
					button.style.transform = "initial";
				}
			});
		} else {
			// default properties
			options.tab.label = options.tab.label || "Need Help?";
			options.tab.color = options.tab.color || "#0b3d91";
			options.tab.fontColor = options.tab.fontColor || "#ffffff";
			options.tab.fontSize = options.tab.fontSize || "16";
			options.tab.size.width = options.tab.size.width || "150";
			options.tab.size.height = options.tab.size.height || "60";
			options.tab.placement.side = options.tab.placement.side || "right";
			options.tab.placement.offset = options.tab.placement.offset || "50";

			var useConfig = {
				setColors: function(el, color, bgColor) {
					el.style.color = color;
					el.style.backgroundColor = bgColor;
				},

				setText: function(el, label, fontSize) {
					var p = document.createElement("p");
					p.append( document.createTextNode(label) );
					if ( fontSize !== "16" ) {
						if ( !isNaN(fontSize) ) {
							el.setAttribute("class", "noImage");
							el.style.fontSize = fontSize + "px";
						} else {
							console.log("Invalid value for font size. Please check the configuration file.");
						}
					}
					el.appendChild(p);
				},

				setDimensions: function(el, width, height) {
					if ( !isNaN(width) && !isNaN(height) ) {
						el.style.width = width + "px";
						el.style.height = height + "px";
					} else {
						if ( isNaN(width) ) {
							console.log("Invalid value for tab WIDTH. Please check the configuration file.");
						}
						if ( isNaN(height) ) {
							console.log("Invalid value for tab HEIGHT. Please check the configuration file.");
						}
					}
				},

				calculateAdjustment: function(width, height) {
					return -0.5 * ( Number(width) - Number(height) ) - 5;
				},

				calculateMaxOffset: function(width, height) {
					return [ window.innerHeight - 0.5 * ( Number(width) + Number(height) ), window.innerWidth - Number(width) ];
				},

				setPlacement: function(el, side, offset, maxOffset, adjustment) {
					if ( !isNaN(offset) ) {
						if ( side === "right" || side === "left" ) {
							var os = Number(offset) * window.innerHeight / 100,
								minOffset = -1 * ( Number(adjustment) + 5 ),
								adjust = ( adjustment !== undefined );
							if ( os < minOffset ) {
								el.style.top = minOffset + "px";
							} else if ( os > Number(maxOffset[0]) ) {
								el.style.top = maxOffset[0] + "px";
							} else {
								el.style.top = offset + "vh";
							}
							if ( side === "right" ) {
								el.setAttribute("class", "feedbackTab");
								if ( adjust ) {
									el.style.right = adjustment + "px";
								}
							} else  {
								el.setAttribute("class", "feedbackTab left");
								if ( adjust ) {
									el.style.left = adjustment + "px";
								}
							}
						} else if (side === "top" || side === "bottom" ) {
							if ( Number(offset) < 0 ) {
								el.style.left = "0";
							} else if ( Number(offset) * window.innerWidth / 100 > Number(maxOffset[1]) ) {
								el.style.left = maxOffset[1] + "px";
							} else {
								el.style.left = offset + "vw";
							}
							if ( side === "top" ) {
								el.setAttribute("class", "feedbackTab top");
							} else {
								el.setAttribute("class", "feedbackTab bottom");
							}
						} else {
							console.log("Invalid value for SIDE of screen to place the tab. The valid options " +
								"are LEFT, RIGHT, TOP, or BOTTOM. Please check the configuration file.");
						}
					} else {
						console.log("Invalid value for OFFSET of tab placement. Please check the configuration file.");
					}
				}
			};

			useConfig.setColors(button, options.tab.fontColor, options.tab.color);
			useConfig.setText(button, options.tab.label, options.tab.fontSize);

			var	adjustment,
				width = Math.max( Number(options.tab.size.width), Number(options.tab.size.height) ),
				height = Math.min( Number(options.tab.size.width), Number(options.tab.size.height) ),
				defaultWidth = ( width === 150 ),
				defaultHeight = ( height === 60 );
			if ( !defaultWidth || !defaultHeight ) {
				useConfig.setDimensions(button, width, height);
				adjustment = useConfig.calculateAdjustment(width, height);
			}

			var side = options.tab.placement.side.toLowerCase(),
				offset = options.tab.placement.offset,
				maxOffset = useConfig.calculateMaxOffset(width, height);
			if ( offset !== "50" || side !== "right" || adjustment !== undefined ) {
				useConfig.setPlacement(button, side, offset, maxOffset, adjustment);
			}
		}

		if ( !button.classList.contains("feedbackTab") ) {
			button.className = "feedbackTab";
		}
		button.onclick = returnMethods.open;

		if ( options.appendTo !== null ) {
			((options.appendTo !== undefined) ? options.appendTo : document.body).appendChild( button );
		}

		return returnMethods;
	};

	window.Feedback.Page = function() {};
	window.Feedback.Page.prototype = {

		render: function( dom ) {
			this.dom = dom;
		},
		start: function() {},
		close: function() {},
		data: function() {
			// don't collect data from page by default
			return false;
		},
		end: function() { return true; }

	};
	window.Feedback.Send = function() {};
	window.Feedback.Send.prototype = {

		send: function() {}

	};

	window.Feedback.Form = function( elements ) {

		this.elements = elements || [
			{
				type: "input",
				id: "feedback-name",
				name: "Name",
				label: "Name",
				required: false
			},
			{
				type: "input",
				id: "feedback-email",
				name: "Email",
				label: "Email",
				required: false
			},
			{
				type: "select",
				id: "feedback-type",
				name: "Type",
				label: "Type",
				values: config.feedback.feedbackType,
				required: false
			},
			{
				type: "textarea",
				id: "feedback-comment",
				name: "Comment",
				label: "Comment",
				required: true
			}
		];
		this.dom = document.createElement("div");
		this.dom.className = "feedback-form-container";

	};

	window.Feedback.Form.prototype = new window.Feedback.Page();

	window.Feedback.Form.prototype.render = function() {

		var i = 0, len = this.elements.length, item;
		emptyElements( this.dom );
		for (; i < len; i++) {
			item = this.elements[ i ];
			var div = document.createElement("div");
			div.classList.add("feedback-input");

			var formEl = document.createElement( item.type );
			formEl.name = item.name;
			formEl.id = item.id;
			
			if ( item.required ) {
				formEl.required = true;
				div.appendChild( element("label", item.label + ": *"));
			} else {
				div.appendChild( element("label", item.label + ":"));
			}

			if (item.type === "select") {
				var options = item.values.split(",");
				for (j = 0; j < options.length; j++) {
					var option = document.createElement("option");
					option.value = option.textContent = options[j];
					formEl.appendChild(option);
				}
			}

			div.appendChild( (item.element = formEl) );
			this.dom.appendChild(div);
		}

		return this;

	};

	window.Feedback.Form.prototype.end = function() {
		// form validation
		var i = 0, len = this.elements.length, item;
		for (; i < len; i++) {
			item = this.elements[ i ];

			// check that all required fields are entered
			if ( item.required === true && item.element.value.length === 0) {
				item.element.className = "feedback-error";
				return false;
			} else {
				item.element.className = "";
			}
		}

		return true;

	};

	window.Feedback.Form.prototype.data = function() {
		var i = 0, len = this.elements.length, item, data = {};

		for (; i < len; i++) {
			item = this.elements[ i ];
			data[ item.name ] = item.element.value;
		}

		// cache and return data
		return ( this._data = data );
	};

	window.Feedback.XHR = function( url ) {

		this.xhr = new XMLHttpRequest();
		this.url = url;

	};

	window.Feedback.XHR.prototype = new window.Feedback.Send();

	window.Feedback.XHR.prototype.send = function( data, callback ) {

		var xhr = this.xhr;

		xhr.onreadystatechange = function() {
			if( xhr.readyState == 4 ) {
				callback( (xhr.status === 200) );
			}
		};

		var emailData = '';
		emailData = 'subject=Feedback from ' + window.location.hostname;
		emailData += '&content=';

		for (var key in data) {
			emailData += key + ': ';
			emailData += data[key] + '\n';
		}

		emailData += '\nLocation: ' + window.location.href + '\n';

		xhr.open( "POST", this.url, true);
		xhr.setRequestHeader('Content-type','application/x-www-form-urlencoded');
		xhr.send(emailData);
	};

})( window, document );

feedback/recaptcha-v3-verify.php
<!-- 
	Copyright (c) 2019, California Institute of Technology ("Caltech").  
	U.S. Government sponsorship acknowledged.

	All rights reserved.

	Redistribution and use in source and binary forms, with or without
	modification, are permitted provided that the following conditions are met:

	* Redistributions of source code must retain the above copyright notice,
	  this list of conditions and the following disclaimer.
	* Redistributions must reproduce the above copyright notice, this list of
	  conditions and the following disclaimer in the documentation and/or other
	  materials provided with the distribution.
	* Neither the name of Caltech nor its operating division, the Jet Propulsion
	  Laboratory, nor the names of its contributors may be used to endorse or
	  promote products derived from this software without specific prior written
	  permission.

	THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
	AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
	IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
	ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
	LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
	CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
	SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
	INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
	CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
	ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
	POSSIBILITY OF SUCH DAMAGE.
-->
<?php

	$url = 'https://www.google.com/recaptcha/api/siteverify';

	// Replace $SECRET_CODE with the Google Recaptcha secret code
	// Contact PDS Engineering Node at pds_operator@jpl.nasa.gov to get this information.
	$secret = '$SECRET_CODE';

	$response = $_POST['response'];
	$verifyURL = $url . '?secret=' . urlencode($secret) . '&response=' . urlencode($response);
	$verifyResponse = file_get_contents($verifyURL);
	$responseData = json_decode($verifyResponse);

	var_dump($responseData);

	if ($responseData && $responseData->action === $action) {
		return $responseData->score;
	} else {
		return false;
	}
?>


feedback/js/config.js
var config = {
	host: "https://pds.nasa.gov",	// Email Server Host
	feedback: {
		header: "Help Desk",
		feedbackType: "Comment,Question,Problem/Bug,Kudos,Other",
		additionalLinks: [
			{
				title: "Information for Data Users",
				url: "https://pds.nasa.gov/home/users/"
			},
			{
				title: "Information for Proposers",
				url: "https://pds.nasa.gov/home/proposers/"
			},
			{
				title: "Information for Providers",
				url: "https://pds.nasa.gov/home/providers/"
			},
			{
				title: "OpenPlanetary Forum",
				url: "https://forum.openplanetary.org/"
			},
			{
				title: "Site Map",
				url: "https://pds.nasa.gov/site-help.shtml"
			}
		]
	},
	tab: {
		label: "Need Help?",		// default: Need Help?
		color: "#0b3d91",			// default: #0b3d91
		fontColor: "#ffffff",		// default: #ffffff
		fontSize: "16",				// default: 16
		size: {
			width: "150",			// default: 150
			height: "60"			// default: 60
		},
		placement: {
			side: "right",			// default: right
			offset: "50"			// default: 50
		}
	}
};
